filebeat/input/tcp/harvester.go

```diff
@@ -0,0 +1,157 @@
+package tcp
+
+import (
+	"bufio"
+	"bytes"
+	"net"
+	"sync"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/elastic/beats/filebeat/harvester"
+)
+
+// Harvester represent a TCP server
+type Harvester struct {
+	sync.RWMutex
+	forwarder *harvester.Forwarder
+	config    config
+	server    net.Listener
+	clients   map[*Client]struct{}
+	wg        sync.WaitGroup
+	done      chan struct{}
+	splitFunc bufio.SplitFunc
+}
+
+// NewHarvester creates a new harvester that will forward events
+func NewHarvester(
+	forwarder *harvester.Forwarder,
+	cfg *common.Config,
+) (*Harvester, error) {
+
+	config := defaultConfig
+	err := cfg.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	server, err := net.Listen("tcp", config.Host)
+	if err != nil {
+		return nil, err
+	}
+
+	sf := splitFunc([]byte(config.LineDelimiter))
+	return &Harvester{
+		config:    config,
+		forwarder: forwarder,
+		clients:   make(map[*Client]struct{}, 0),
+		done:      make(chan struct{}),
+		server:    server,
+		splitFunc: sf,
+	}, nil
+}
+
+// Run start and run a new TCP listener to receive new data
+func (h *Harvester) Run() error {
+	logp.Info("Started listening for incoming TCP connection on: %s:", h.config.Host)
+	for {
+		conn, err := h.server.Accept()
+		if err != nil {
+			select {
+			case <-h.done:
+				return nil
+			default:
+				logp.Debug("tcp", "Can not accept the connection", err)
+				continue
+			}
+		}
+
+		client := NewClient(
+			conn,
+			h.forwarder,
+			h.splitFunc,
+			h.config.MaxMessageSize,
+			h.config.Timeout,
+		)
+		logp.Debug(
+			"tcp",
+			"New client, remote: %s (total clients: %d)",
+			conn.RemoteAddr(),
+			h.clientsCount(),
+		)
+		h.wg.Add(1)
+		go func() {
+			defer conn.Close()
+
+			h.registerClient(client)
+			err := client.Handle()
+			if err != nil {
+				logp.Debug("tcp", "Client error", err)
+			}
+			h.unregisterClient(client)
+			logp.Debug(
+				"tcp",
+				"Client disconnected, remote: %s (total clients: %d)",
+				conn.RemoteAddr(),
+				h.clientsCount(),
+			)
+			h.wg.Done()
+		}()
+	}
+}
+
+// Stop stops accepting new incoming TCP connection and close any active clients
+func (h *Harvester) Stop() {
+	logp.Info("Stopping TCP harvester")
+	close(h.done)
+	h.server.Close()
+
+	logp.Debug("tcp", "Closing remote connections")
+	for _, client := range h.allClients() {
```

Comment:
I think you could eliminate calling `allClients` and simply loop over the map:
```
for client, _ := range h.clients {
    client.Close()
}
```
At this point there can't be any new connection, right? Does locking make sense here? If locking is required, I am ok with keeping the function. But the transformation seems unnecessary.

Reply:
When this code is run we stop refusing new connection but, when the client is shutting down each client are removed from the hash at the end of the [goroutine](https://github.com/elastic/beats/pull/6266/files#diff-d84dc93dff4114566dc2663822cc67adR92). 

So I have two options either I use a copy of the original hash which is what *allClients* is doing or I lock before looping the element in the array this will make all the goroutine wait for the mutex before they can remove themself. I went with the former to keep consistency in the values in the hash during the loop.